test/e2e/upgrade_test.go

```diff
@@ -0,0 +1,131 @@
+}
+
+// testUpgradeVersion tests if the operator deletes machines and recreats them, if the node annotation doesn`t match
+// with the expected annotation
+func testUpgradeVersion(t *testing.T) {
+	testCtx, err := NewTestContext(t)
+	require.NoError(t, err)
+
+	operatorVersion, err := testCtx.getWMCOVersion()
+	require.NoError(t, err, "could not get WMCO version")
+
+	err = testCtx.waitForWindowsNodesAnnotation(gc.numberOfNodes, operatorVersion)
+	if err != nil {
+		t.Fatalf("windows node upgrade failed  with %v", err)
+	}
+
+	// Get all the Windows nodes
+	nodes, err := framework.Global.KubeClient.CoreV1().Nodes().List(context.TODO(),
+		metav1.ListOptions{LabelSelector: nc.WindowsOSLabel})
+	require.NoError(t, err)
+
+	for _, node := range nodes.Items {
+		t.Run(node.GetName(), func(t *testing.T) {
+			require.Containsf(t, node.Annotations, nc.VersionAnnotation, "node %s missing version annotation", node.GetName())
+			assert.Equal(t, operatorVersion, node.Annotations[nc.VersionAnnotation], "Node recreation failed. WMCO version annotation mismatch")
+		})
+	}
+}
+
+// testTamperAnnotation tests if the operator deletes machines and recreats them, if the node annotation is changed to an invalid value
+// with the expected annotation
+func testTamperAnnotation(t *testing.T) {
+	testCtx, err := NewTestContext(t)
+	require.NoError(t, err)
+
+	operatorVersion, err := testCtx.getWMCOVersion()
+	require.NoError(t, err, "could not get WMCO version")
+
+	// tamper node annotation
+	nodes, err := framework.Global.KubeClient.CoreV1().Nodes().List(context.TODO(),
+		metav1.ListOptions{LabelSelector: nc.WindowsOSLabel})
+	require.NoError(t, err)
+
+	for _, node := range nodes.Items {
+		patchData := fmt.Sprintf(`{"metadata":{"annotations":{"%s":"%s"}}}`, nc.VersionAnnotation, "0.0.2+baa8c1f-dirty")
+		_, err := framework.Global.KubeClient.CoreV1().Nodes().Patch(context.TODO(), node.Name, types.MergePatchType, []byte(patchData), metav1.PatchOptions{})
+		require.NoError(t, err)
+		if err == nil {
+			break
+		}
+	}
+
+	err = testCtx.waitForWindowsNodesAnnotation(1, operatorVersion)
+	if err != nil {
+		t.Fatalf("windows node upgrade failed  with %v", err)
+	}
+
+	// Get all the Windows nodes
+	nodes, err = framework.Global.KubeClient.CoreV1().Nodes().List(context.TODO(),
+		metav1.ListOptions{LabelSelector: nc.WindowsOSLabel})
+	require.NoError(t, err)
+
+	for _, node := range nodes.Items {
+		t.Run(node.GetName(), func(t *testing.T) {
+			require.Containsf(t, node.Annotations, nc.VersionAnnotation, "node %s missing version annotation", node.GetName())
+			assert.Equal(t, operatorVersion, node.Annotations[nc.VersionAnnotation], "Node recreation failed. WMCO version annotation mismatch")
+		})
+	}
+}
+
+// waitForWindowsNodesAnnotation waits until the given number of nodes are created with the expected version annotation
+func (tc *testContext) waitForWindowsNodesAnnotation(nodeCount int32, operatorVersion string) error {
```

Comment:
Yes, I did not use `waitForWindowsNodes` because it takes time for nodes to be deleted and due to that `waitForWindowsNodes` will immediately retrurn true. To avoid that, I had to add [this](https://github.com/openshift/windows-machine-config-operator/pull/101/files#diff-225c6dc1116ed17311dbd31aef8e09e9R119) condition and hence needed to create this function.

Reply:
Is there a way to share the common code by adding a parameter input? Only if that parameter is true will it check `if currAnn != operatorVersion`